src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -808,6 +808,11 @@ public RequestLineStatus TakeStartLine(SocketInput input)
 
                 ConnectionControl.CancelTimeout();
 
+                if (scan.Peek() == 0x16)
+                {
+                    throw new IOException("SSL/TLS handshake was attempted on an HTTP endpoint.");
```

Comment:
Since we're just checking the first byte, and a client could send 0x16 for any reason (though most likely it was TLS handshake), I would change the error message to say it *might* have been an SSL/TLS handshake.